[IMP] barcode: update default nomenclature doc (#8044)
odoo/documentation: content/applications/inventory_and_mrp/barcode/operations/barcode_nomenclature.rst

============================
Default barcode nomenclature
============================

.. |UPC| replace:: :abbr:`UPC (Universal Product Code)`
.. |EAN| replace:: :abbr:`EAN (European Article Number)`
.. |GTIN| replace:: :abbr:`GTIN (Global Trade Item Number)`
.. |GTINs| replace:: :abbr:`GTINs (Global Trade Item Numbers)`

*Barcode nomenclatures* define how barcodes are recognized and categorized. When a barcode is
scanned, it is associated to the **first** rule with a matching pattern. The pattern syntax is
described in Odoo's nomenclature list using a regular expression, and a barcode is successfully read
by Odoo if its prefix and/or length matches the one defined in the barcode's rule.

Odoo *Barcode* supports UPC (Universal Product Code), EAN (European Article Number), and GS1
encoding. The pre-configured nomenclatures in Odoo are *Default Nomenclature* and *Default GS1
Nomenclature*. The default nomenclature uses |UPC| and |EAN| encoding, and supports |UPC|/|EAN|
conversion.

.. important::
   |UPC| and |EAN| barcodes **must** be `purchased from GS1
   <https://www.gs1.org/standards/get-barcodes>`_ to use these barcodes. GS1 is the **only**

>    official provider of |UPC|/|EAN| barcodes globally (as well as GS1 |GTINs|).

Author: Suggestion implemented ✅

.. _barcode/operations/set-up-nomenclature:

Set up barcode nomenclature
===========================

To use default nomenclature, navigate to :menuselection:`Inventory app --> Configuration -->
Settings`. Under the :guilabel:`Barcode` section, click the checkbox next to :guilabel:`Barcode
Scanner` to enable barcodes. Doing so installs the *Barcode* app in the database.

Next, ensure that the :guilabel:`Default Nomenclature` setting is selected in the :guilabel:`Barcode

> Nomenclature` field.

Author: Suggestion implemented ✅

.. image:: barcode_nomenclature/barcode-nomenclature-enabled-setting.png
   :align: center
   :alt: Enabled barcode setting with Default Nomenclature selected.

Once the nomenclature is selected, the :menuselection:`Barcode Nomenclatures` settings can be
accessed by a hidden menu that's discoverable after enabling :ref:`developer mode <developer-mode>`.
Once enabled, navigate to :menuselection:`Inventory app --> Configuration --> Barcode Nomenclatures`

> menu, and select :guilabel:`Default Nomenclature`.

Author: I don't want to add a screenshot of the whole page, because that's what the table at the bottom of the doc is for. So, as a compromise, I added in a ref link to the table at the bottom in the paragraph that closes this section. Then, I added in a screenshot of the fields at the top of the page that are explained in this section.

> menu, and select :guilabel:`Default Nomenclature`.

Author: Suggestion implemented ✅

From this page, the list of *rules* and *barcode patterns* Odoo supports for the default
nomenclature is accessible. The table contains all the information that can be condensed with a

> |UPC|/|EAN| barcode, along with the corresponding :guilabel:`Barcode Pattern`.

Author: Added in an explanation for each of those fields. Suggestion implemented ✅

.. _barcode/operations/use-upc-ean-barcodes:

Use UPC/EAN barcodes in Odoo
============================

For product identification using |UPC|/|EAN| barcodes in Odoo, businesses obtain `barcodes
<https://www.gs1us.org/upcs-barcodes-prefixes/how-to-get-a-upc-barcode>`_ purchased directly from

> GS1.

Author: Suggestion implemented ✅

The |UPC| and |EAN| barcode formats are predominantly used in their own regions. |UPC| is used only
in the United States and Canada, while |EAN| is used everywhere else globally.

A |UPC| is usually a 12-digit barcode used to identify most products, while |EAN| barcodes use
13-digit barcodes to identify products.

|UPC| codes can be converted to |EAN| by prefixing them with a zero. In Odoo, by default,
|UPC|/|EAN| barcodes are automatically converted one way or another when trying to match a rule with
the other encoding.

Refer to the :ref:`default nomenclature list <barcode/operations/default-nomenclature-list>` to see
a comprehensive list of all barcode patterns and rules to follow.

Create rules
------------

|UPC| and |EAN| barcodes contain information in the barcode. Scanning these codes from the
:ref:`default nomenclature list <barcode/operations/default-nomenclature-list>` automatically fills

> corresponding data in the Odoo database.

Author: Suggestion implemented ✅

Adding new barcode rules to this list ensures that non-standard (:dfn:`user-created`) formats are
interpreted accurately.

To create new rules, first activate :ref:`developer mode <developer-mode>`. Then, navigate to
:menuselection:`Inventory app --> Configuration --> Barcode Nomenclatures`. Select the
:guilabel:`Default Nomenclature` list item.

On the :guilabel:`Default Nomenclature` page, select :guilabel:`Add a line` at the bottom of the
table, which opens a window to create a new rule.

.. image:: barcode_nomenclature/barcode-nomenclature-new-rule-popup.png
   :align: center
   :alt: Create Rules pop-up window on Default Nomenclature page.

The :guilabel:`Rule Name` field is used internally to identify what the barcode represents. The
barcode :guilabel:`Types` are different classifications of information that can be understood by the
system (e.g., Package, Lot, Location, Coupon, etc.). The :guilabel:`Sequence` represents the

> priority of the rule; this means the smaller the value, the higher the rule appears on the table.

Reviewer: Do you want to add a quick mention/explanation of the 'Encoding' field, as well? Just to complete the pop-up window field explanations?

> Odoo follows the sequential order of this table, and will use the first rule it matches based on the
> sequence. The :guilabel:`Barcode Pattern` is how the sequence of letters or numbers is recognized by
> the system to contain information about the product.

Reviewer: ```suggestion
The :guilabel:`Barcode Pattern` represents how the sequence of letters or numbers is recognized by
the system to contain information about the product. Odoo follows the sequential order of this
table, and uses the first rule it matches, based on the sequence. 
```

Reviewer: just to follow the order of things as they appear on the pop-up

Author: Suggestion implemented ✅

.. note::
   Barcode patterns can also define how numerical values, such as weight or price, are encoded in
   the barcode.

>    They are indicated by **{NNN}**, where N defines where the number's digits are encoded. *Floats*
>    are also supported with the decimals, indicated by D, such as **{NNNDD}**.

Reviewer: ```suggestion
   They are indicated by **{NNN}**, where 'N' defines where the number's digits are encoded. *Floats*
   are also supported with the decimals, indicated by 'D', such as **{NNNDD}**.
```

Author: Suggestion implemented ✅

>    In this case, the barcode field on the associated records **must** show these digits as zeroes.

Reviewer: ```suggestion
   In this case, the barcode field on the associated records **must** show these digits as zeros.
```

Author: Suggestion implemented ✅

> After filling in the information, click the :guilabel:`Save & New` button to make another rule, or
> click :guilabel:`Save & Close` to save and return to the table of rules.

Reviewer: ```suggestion
After filling in the information, click the :guilabel:`Save & New` button to save the rule, and
instantly start creating another rule. Or, click :guilabel:`Save & Close` to save the rule, and
return to the table of rules.
```

Author: Suggestion implemented ✅

.. tip::

>    When the :guilabel:`Barcode Pattern` field contains `.*`, it means that it can contain **any**
>    number of characters, those characters being **any** number.

Reviewer: ```suggestion
   When the :guilabel:`Barcode Pattern` field contains `.*`, it means it can contain **any**
   number of characters - those characters being **any** number.
```

.. _barcode/operations/default-nomenclature-list:

Default nomenclature list
=========================

> The table below contains Odoo's list of default nomenclature rules. Barcode patterns are written in

Reviewer: ```suggestion
The table below contains Odoo's list of Default Nomenclature rules. Barcode patterns are written in
```

barcode.

Odoo *Barcode* supports UPC (Universal Product Code), EAN (European Article Number), and GS1
encoding. The pre-configured nomenclatures in Odoo are *Default Nomenclature* and *Default GS1
Nomenclature*. The default nomenclature uses |UPC| and |EAN| encoding, and supports |UPC|/|EAN|
conversion.

.. important::
   |UPC| and |EAN| barcodes **must** be `purchased from GS1
   <https://www.gs1.org/standards/get-barcodes>`_ to use these barcodes. GS1 is the **only**
   official provider of |UPC|/|EAN| barcodes globally (as well as GS1 |GTINs|).

.. _barcode/operations/set-up-nomenclature:

Set up barcode nomenclature
===========================

To use default nomenclature, navigate to :menuselection:`Inventory app --> Configuration -->
Settings`. Under the :guilabel:`Barcode` section, click the checkbox next to :guilabel:`Barcode
Scanner` to enable barcodes. Doing so installs the *Barcode* app in the database.

Next, ensure that the :guilabel:`Default Nomenclature` setting is selected in the :guilabel:`Barcode
Nomenclature` field.

.. image:: barcode_nomenclature/barcode-nomenclature-enabled-setting.png
   :align: center

>    :alt: Enabled barcode setting with Default Nomenclature selected.

Author: Thank you!

The list of *rules* and *barcode patterns* Odoo supports for the default nomenclature is accessible
by clicking the :guilabel:`➡️ (arrow)` icon to the right of the :guilabel:`Barcode Nomenclature`

> selection.

Author: I'm seeing the same thing now. I've updated the two paragraphs including this instruction and the one from your comment below to take out this part and instead instruct them to access it the other way.

Suggestion implemented ✅

In the :guilabel:`Open: Nomenclature` pop-up table, view and edit the default :guilabel:`Rule Names`
available in Odoo. The table contains all the information that can be condensed with a |UPC|/|EAN|

> barcode, along with the corresponding :guilabel:`Barcode Pattern`.

Author: This, along with the above comment, has been updated to have the reader access the page a different way, since this one seems to not work in 17.0.

Suggestion implemented ✅

============================
Default barcode nomenclature
============================

.. |UPC| replace:: :abbr:`UPC (Universal Product Code)`
.. |EAN| replace:: :abbr:`EAN (European Article Number)`
.. |GTIN| replace:: :abbr:`GTIN (Global Trade Item Number)`
.. |GTINs| replace:: :abbr:`GTINs (Global Trade Item Numbers)`

*Barcode nomenclatures* define how barcodes are recognized and categorized. When a barcode is
scanned, it is associated to the **first** rule with a matching pattern. The pattern syntax is
described in Odoo's nomenclature list using a regular expression, and a barcode is successfully read
by Odoo if its prefix and/or length matches the one defined in the barcode's rule.

Odoo *Barcode* supports UPC (Universal Product Code), EAN (European Article Number), and GS1
encoding. The pre-configured nomenclatures in Odoo are *Default Nomenclature* and *Default GS1
Nomenclature*. The default nomenclature uses |UPC| and |EAN| encoding, and supports |UPC|/|EAN|
conversion.

.. important::
   |UPC| and |EAN| barcodes **must** be `purchased from GS1
   <https://www.gs1.org/standards/get-barcodes>`_ to use these barcodes. GS1 is the **only**

>    official provider of |UPC|/|EAN| and GS1 |GTINs| globally.

Reviewer: ```suggestion
   official provider of |UPC|/|EAN| and GS1 |GTINs| in the world.
```
just sounds a bit cleaner with this phrasing

Author: Suggestion implemented ✅

Settings`. Under the :guilabel:`Barcode` section, click the checkbox next to :guilabel:`Barcode
Scanner` to enable barcodes. Doing so installs the *Barcode* app in the database.

Next, ensure that the :guilabel:`Default Nomenclature` setting is selected in the :guilabel:`Barcode
Nomenclature` field. Then, click :guilabel:`Save`.

.. image:: barcode_nomenclature/barcode-nomenclature-enabled-setting.png
   :align: center
   :alt: Enabled barcode setting with Default Nomenclature selected.

Once the nomenclature is saved and selected, the :menuselection:`Barcode Nomenclatures` settings can
be accessed, via a hidden menu that is **only** discoverable after enabling :ref:`developer mode
<developer-mode>`.

Once developer mode is enabled, navigate to :menuselection:`Inventory app --> Configuration -->
Barcode Nomenclatures` menu, and select :guilabel:`Default Nomenclature`.

From this page, the :guilabel:`Barcode Nomenclature` is specified at the top as the `Default
Nomenclature`.

Under that, the :guilabel:`|UPC|/|EAN| Conversion` field is set to :guilabel:`Always`, by default.
This setting determines if a |UPC|/|EAN| barcode should be automatically converted one way or

> another when trying to match a rule with the other encoding. The other available options for this
> field are :guilabel:`Never`, :guilabel:`EAN-13 to UPC-A`, and :guilabel:`UPC-A to EAN-13`.

Reviewer: ```suggestion
another when trying to match a rule with the other encoding.

The other available options for this field are: :guilabel:`Never`, :guilabel:`EAN-13 to UPC-A`, and
:guilabel:`UPC-A to EAN-13`.
```

Author: Suggestion implemented ✅

.. important::
   For |UPC|/|EAN| conversion to work for every barcode scanned, the setting in the
   :guilabel:`|UPC|/|EAN| Conversion` field **must** be set to :guilabel:`Always`.

> The last field at the top of the page is the :guilabel:`Is GS1 Nomenclature` field. For the Default
> Nomenclature, this field should remain unchecked. This is because the default nomenclature uses
> |UPC| and |EAN| encoding, *not* GS1 encoding.

Reviewer: ```suggestion
The last field at the top of the page is the :guilabel:`Is GS1 Nomenclature` field. For the
:guilabel:`Default Nomenclature`, this field should remain unchecked. This is because the
:guilabel:`Default Nomenclature` uses |UPC| and |EAN| encoding, *not* GS1 encoding.
```

Author: Suggestion implemented ✅

a comprehensive list of all barcode patterns and rules to follow.

Create rules
------------

|UPC| and |EAN| barcodes contain specific information in the barcode. Scanning these codes from the
:ref:`default nomenclature list <barcode/operations/default-nomenclature-list>` automatically fills
corresponding data in the Odoo database.

Adding new barcode rules to this list ensures that non-standard (:dfn:`user-created`) formats are
interpreted accurately.

To create new rules, first activate :ref:`developer mode <developer-mode>`. Then, navigate to
:menuselection:`Inventory app --> Configuration --> Barcode Nomenclatures`. Select the
:guilabel:`Default Nomenclature` list item.

On the :guilabel:`Default Nomenclature` page, select :guilabel:`Add a line` at the bottom of the
table, which opens a :guilabel:`Create Rules` pop-up window to create a new rule.

.. image:: barcode_nomenclature/barcode-nomenclature-new-rule-popup.png
   :align: center
   :alt: Create Rules pop-up window on Default Nomenclature page.

The :guilabel:`Rule Name` field is used internally to identify what the barcode represents.

> The :guilabel:`Sequence` represents the priority of the rule; meaning the smaller the value, the

Reviewer: ```suggestion
The :guilabel:`Sequence` field represents the priority of the rule; meaning the smaller the value, the
```

Author: Suggestion implemented ✅

higher the rule appears on the table.

> The barcode :guilabel:`Type` represents different classifications of information that can be

Reviewer: ```suggestion
The barcode :guilabel:`Type` field represents different classifications of information that can be
```

Author: Suggestion implemented ✅

understood by the system (e.g., :guilabel:`Package`, :guilabel:`Lot`, :guilabel:`Location`,
:guilabel:`Coupon`, etc.).

> The :guilabel:`Encoding` specifies which encoding the barcode uses; this rule will only apply if the

Reviewer: ```suggestion
The :guilabel:`Encoding` field specifies which encoding the barcode uses; this rule **only** applies if the
```

Author: Suggestion implemented ✅

> barcode uses this specific encoding. The available :guilabel:`Encoding` options are **EAN-13**,

Reviewer: ```suggestion
barcode uses this specific encoding. The available :guilabel:`Encoding` options are: **EAN-13**,
```
if these options appear on the UI (if chosen), wouldn't a GUI label be used in this instance? What are your thoughts?

Author: I went back and forth between the two, but yeah I agree with you - so I've made the change. Suggestion implemented ✅

**EAN-8**, **UPC-A**, and **GS1-28**.

> The :guilabel:`Barcode Pattern` represents how the sequence of letters or numbers is recognized by

Reviewer: ```suggestion
The :guilabel:`Barcode Pattern` field represents how the sequence of letters or numbers is recognized by
```

Author: Suggestion implemented ✅

the system to contain information about the product. Odoo follows the sequential order of this
table, and uses the first rule it matches, based on the sequence.

.. note::
   Barcode patterns can also define how numerical values, such as weight or price, are encoded in
   the barcode.

   They are indicated by **{NNN}**, where `N` defines where the number's digits are encoded.
   *Floats* are also supported with the decimals, indicated by `D`, such as **{NNNDD}**.

   In this case, the barcode field on the associated records **must** show these digits as zeros.

After filling in the information, click the :guilabel:`Save & New` button to save the rule, and
instantly start creating another rule. Or, click :guilabel:`Save & Close` to save the rule, and
return to the table of rules.

.. tip::
   When the :guilabel:`Barcode Pattern` field contains `.*`, it means it can contain **any**
   number of characters - those characters being **any** number.

.. _barcode/operations/default-nomenclature-list:

Default nomenclature list
=========================

> The table below contains Odoo's list of Default Nomenclature rules. Barcode patterns are written in

Reviewer: ```suggestion
The table below contains Odoo's list of :guilabel:`Default Nomenclature` rules. Barcode patterns are written in
```

Author: Suggestion implemented ✅

another when trying to match a rule with the other encoding.

The other available options for this field are :guilabel:`Never`, :guilabel:`EAN-13 to UPC-A`, and
:guilabel:`UPC-A to EAN-13`.

.. important::
   For |UPC|/|EAN| conversion to work for every barcode scanned, the setting in the
   :guilabel:`|UPC|/|EAN| Conversion` field **must** be set to :guilabel:`Always`.

The last field at the top of the page is the :guilabel:`Is GS1 Nomenclature` field. For the
:guilabel:`Default Nomenclature`, this field should remain unchecked. This is because the
:guilabel:`Default Nomenclature` uses |UPC| and |EAN| encoding, *not* GS1 encoding.

.. image:: barcode_nomenclature/barcode-nomenclature-page-fields.png
   :align: center
   :alt: Default Nomenclature page setting fields.

Further down the page, there is a list showcasing the :guilabel:`Rule Name`, :guilabel:`Type`,
:guilabel:`Encoding`, and :guilabel:`Barcode Pattern` for the *rules* and *barcode patterns* that
Odoo supports for the default nomenclature.

The :ref:`table <barcode/operations/default-nomenclature-list>` contains all the information that

> can be condensed with a |UPC|/|EAN| barcode.

Author: Suggestion implemented ✅

if the barcode uses this specific encoding. The available :guilabel:`Encoding` options are:
:guilabel:`EAN-13`, :guilabel:`EAN-8`, :guilabel:`UPC-A`, and :guilabel:`GS1-28`.

The :guilabel:`Barcode Pattern` field represents how the sequence of letters or numbers is
recognized by the system to contain information about the product. Odoo follows the sequential order
of this table, and uses the first rule it matches, based on the sequence.

.. note::
   Barcode patterns can also define how numerical values, such as weight or price, are encoded in
   the barcode.

   They are indicated by **{NNN}**, where `N` defines where the number's digits are encoded.
   *Floats* are also supported with the decimals, indicated by `D`, such as **{NNNDD}**.

   In this case, the barcode field on the associated records **must** show these digits as zeros.

After filling in the information, click the :guilabel:`Save & New` button to save the rule, and
instantly start creating another rule. Or, click :guilabel:`Save & Close` to save the rule, and
return to the table of rules.

.. tip::
   When the :guilabel:`Barcode Pattern` field contains `.*`, it means it can contain **any**

>    number of characters - those characters being **any** number.

Author: Suggestion implemented ✅

>    number of characters - those characters being **any** number.

Author: Ah, that's interesting thank you for pointing that out! 🙏

.. _barcode/operations/default-nomenclature-list:

Default nomenclature list
=========================

The table below contains Odoo's list of :guilabel:`Default Nomenclature` rules. Barcode patterns are
written in regular expressions.

.. list-table::
   :header-rows: 1
   :stub-columns: 1

   * - Rule Name
     - Type
     - Encoding
     - Barcode Pattern
   * - Price Barcodes 2 Decimals
     - Priced Product
     - EAN-13
     - 23.....{NNNDD}
   * - Discount Barcodes
     - Discounted Product
     - Any
     - 22{NN}
   * - Weight Barcodes 3 Decimals
     - Weighted Product
     - EAN-13
     - 21.....{NNDDD}
   * - Customer Barcodes
     - Client
     - Any
     - 042
   * - Coupon & Gift Card Barcodes
     - Coupon
     - Any
     - 043|044
   * - Cashier Barcodes
     - Cashier
     - Any
     - 041
   * - Location barcodes
     - Location
     - Any
     - 414
   * - Package barcodes
     - Package
     - Any
     - PACK
   * - Lot barcodes
     - Lot
     - Any
     - 10
   * - Magnetic Credit Card
     - Credit Card
     - Any
     - %.*
   * - Product Barcodes
     - Unit Product
     - Any

>      - .*

Author: Good idea! Suggestion implemented ✅

.. _barcode/operations/set-up-nomenclature:

Set up barcode nomenclature
===========================

To use default nomenclature, navigate to :menuselection:`Inventory app --> Configuration -->
Settings`. Under the :guilabel:`Barcode` section, click the checkbox next to :guilabel:`Barcode
Scanner` to enable barcodes. Doing so installs the *Barcode* app in the database.

Next, ensure that the :guilabel:`Default Nomenclature` setting is selected in the :guilabel:`Barcode
Nomenclature` field. Then, click :guilabel:`Save`.

.. image:: barcode_nomenclature/barcode-nomenclature-enabled-setting.png
   :align: center
   :alt: Enabled barcode setting with Default Nomenclature selected.

Once the nomenclature is saved and selected, the :menuselection:`Barcode Nomenclatures` settings can
be accessed, via a hidden menu that is **only** discoverable after enabling :ref:`developer mode
<developer-mode>`.

Once developer mode is enabled, navigate to :menuselection:`Inventory app --> Configuration -->
Barcode Nomenclatures` menu, and select :guilabel:`Default Nomenclature`.

From this page, the :guilabel:`Barcode Nomenclature` is specified at the top as the `Default
Nomenclature`.

> Under that, the :guilabel:`|UPC|/|EAN| Conversion` field is set to :guilabel:`Always`, by default.

Reviewer 2: The formatting broke here. I suggest to remove the vertical bar symbols `|` from the gui

Author: Removed vertical bar symbols - Suggestion implemented ✅

This setting determines if a |UPC|/|EAN| barcode should be automatically converted one way or
another when trying to match a rule with the other encoding.

The other available options for this field are :guilabel:`Never`, :guilabel:`EAN-13 to UPC-A`, and
:guilabel:`UPC-A to EAN-13`.

.. important::
   For |UPC|/|EAN| conversion to work for every barcode scanned, the setting in the

>    :guilabel:`|UPC|/|EAN| Conversion` field **must** be set to :guilabel:`Always`.

Reviewer 2: Same here, the rendered HTML is not picking up the shortcode markup due to the back ticks in the gui.

Author: Removed vertical bar symbols - Suggestion implemented ✅
